Fix batch updates to allow cleared values to be re-set

grokys opened the pull request:
## What does the pull request do?

@MarchingCube reported a problem introduced by #5070 whereby if a property was cleared during a batch update and then re-set in the `PropertyChanged` notification raised on ending the batch update, an exception would be thrown.

This was caused by our usage of sentinel values to indicate that a local value property entry needs to be removed at the end of a batch update. The fixes are:

- If a new local value is written to the store in the course of ending the batch update, don't update the sentinel value as the value will subsequently be lost. Instead create a new local value
- Allow adding a binding to a cleared property while ending a batch update
- When a binding is added during the end of a batch update, `_batchUpdate` will be non-null but newly added bindings shouldn't have `BeginBatchUpdate` called on them because no `EndBatchUpdate` will arrive (as we've already called them)

Update: it seems #5697 was also caused by batch updating. The `ValueStore` fix is:

- Fix bindings being disposed during batch update

But it seems that batch updates also exposed some problems with transitions. Fix those:

- Handle changing a `Transitions` collections to new collection which contains some of the same elements
- Handle changing transitions during batch update: transition state may not have been initialized by the time the change notification is received for the property which has a transition

Fixes #5697

grokys committed:
Added failing batch update test.

grokys committed:
Don't write new values to remove sentinels.

During a batch update sentinel values are written to the value store to indicate that the value needs to be removed at the end of the update. If a new value is written to the store in the course of ending the batch update, don't update this sentinel value as the value will subsequently be lost.

grokys committed:
Added another failing batch update test.

And a bit of a sanity check to the previous one.

grokys committed:
Allow binding property during ending batch update.

- Allow adding a binding to a cleared property while ending a batch update. Need to check that the existing value isn't a remove sentinel here, otherwise the binding will be lost.
- When a binding is added during the end of a batch update, `_batchUpdate` will be non-null but newly added bindings shouldn't have `BeginBatchUpdate` called on them because no `EndBatchUpdate` will arrive (as we've already called them)
- Add sanity checks to the unit test to make sure that correct notifications are raised

MarchingCube committed:
Merge branch 'master' into fixes/batchupdate-set-cleared-value-end-update

MarchingCube commented on src/Avalonia.Base/ValueStore.cs:
```diff
@@ -103,7 +103,7 @@ public bool IsSet(AvaloniaProperty property)
             IDisposable? result = null;
 
-            if (_values.TryGetValue(property, out var slot))
+            if (_values.TryGetValue(property, out var slot) && !IsRemoveSentinel<T>(slot))
```
I'm still not 100% sure why only certain places respect sentinel values, perhaps we should introduce a small helper to get values (with or without sentinel value handling). And perhaps a comment explaining why we need to handle sentinel values in certain cases would help with maintenance.

grokys replied:
Good point, I've made handling sentinel values more consistent.

MarchingCube commented:
I've asked @Michron to check if it fixes problems on our side.

Commit:
Merge branch 'master' into fixes/batchupdate-set-cleared-value-end-update

MarchingCube reviewed (approved):
Fix confirmed.

grokys committed:
Fix nullable warning.

grokys committed:
Added more failing tests.

To test completing/disposing bindings during and while ending batch updates.

grokys committed:
Fix more batch update problems.

Fixes tests in previous commit which test scenarios where bindings are completed/disposed during and when ending batch updates.

grokys committed:
Added failing transitions test.

grokys committed:
Correctly handle changing transitions.

Handle changing a `Transitions` collections to new collection which contains some of the same elements.

Added a comment explaining why we add new transitions before replacing old transitions; was explained in the commit message for 0694b22c51b96fb8d997ab2dc25a4d5503789a3f.

grokys committed:
Added another failing transitions test.

grokys committed:
Handle changing transitions during batch update.

Fixes failing test from previous commit.

grokys committed:
Make handling of sentinel values consistent.

grokys commented:
@MarchingCube should now also fix #5697.

MarchingCube commented on src/Avalonia.Base/ValueStore.cs:
```diff
@@ -437,9 +456,14 @@ public bool End()
+                            if (values.TryGetValue(entry.property, out var updatedSlot) &&
+                                updatedSlot.Priority == BindingPriority.Unset)
                             {
+                                if (entry.property.Name == "Transitions" && _owner._owner.GetType().Name == "TabItem")
```
Debugging leftover?

grokys replied:
Yep :) removed.

grokys committed:
Remove debugging code.

MarchingCube commented:
Tested again, works fine and fixes both issues 👍

MarchingCube committed:
Merge branch 'master' into fixes/batchupdate-set-cleared-value-end-update
